### ISSUE TITLE
Possible flaw in `Servo.detach()` and other HW timer dependent APIs

### ISSUE THREAD

**Issue by ScruffR** [spark/firmware#745]
Since several servo pins share one hardware timer, I'd see some problem in this implementation of `Servo.detach()`

```cpp
void HAL_Servo_Detach(uint16_t pin)
{
    // TIM disable counter
    STM32_Pin_Info* PIN_MAP = HAL_Pin_Map();
    TIM_Cmd(PIN_MAP[pin].timer_peripheral, DISABLE);
}
```

If there are still other servos depending on this timer, instead of disabling the `timer_peripheral` rather only mask the `timer_ch`.
Only the last "user" of a specific timer should actually be allowed to disable the timer.

But this might call for a centralized way to request access to a HW timer and release it with some kind of reference counter.
  
The OP of following thread also reports issues with re-attaching detached servos
 https://community.particle.io/t/interrupt-and-servos-can-they-work-together/17706

**Comment by m-mcgowan** [comment]
Thanks for raising this. We will investigate in more detail.
